File: python/0225.py
# Link: https://leetcode.com/problems/implement-stack-using-queues/description/
# Runtime: 38 ms
# Memory: 16.38 MB

import queue


class MyStack:
    def __init__(self):
        self.stack_imitation = queue.Queue()
        self.peek = 0

    def push(self, x: int) -> None:
        self.stack_imitation.put(x)
        self.peek = x

    def pop(self) -> int:
        size = self.stack_imitation.qsize()
        while size > 1:
            self.peek = self.stack_imitation.get()
            self.stack_imitation.put(self.peek)
            size -= 1

        top_value = self.stack_imitation.get()

        return top_value

    def top(self) -> int:
        return self.peek

    def empty(self) -> bool:
        return self.stack_imitation.empty()


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: c++/0225.cpp
// Link: https://leetcode.com/problems/implement-stack-using-queues/
// Runtime: 0 ms
// Memory: 6.8 MB

class MyStack {
public:
  MyStack() {}

  void push(int x) {
    peek = x;
    container.push(x);
  }

  int pop() {
    int size = container.size() - 1;
    while (size > 0) {
      peek = container.front();
      container.push(peek);
      container.pop();

      size--;
    }
    int topElement = container.front();
    container.pop();

    return topElement;
  }

  int top() { return peek; }

  bool empty() { return container.empty(); }

private:
  std::queue<int> container;
  int peek = 0;
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */